### DISCUSSION TITLE
Compound Exercise - Microservices - Design Patterns - API Gateway with Aggregator

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#158] [Tasks]
# Spring Microservices Exercise: API Gateway & Aggregator Pattern

## Introduction

In this exercise, you'll build a simplified e-commerce system consisting of multiple microservices. It involves the Customer Service, Product Service, and Order Service. These will be orchestrated by an Aggregator and all the traffic will be routed through an API Gateway.

## Architecture Overview

1. **API Gateway**: Entry point for client requests. It routes the client requests to appropriate microservices.
2. **Aggregator**: A service that talks to other microservices and combines the data.
3. **Customer Service**: Manages customer data.
4. **Product Service**: Manages product data.
5. **Order Service**: Manages order data.

## Prerequisites

- Java 11
- Maven
- IDE of your choice (Eclipse, IntelliJ, etc.)
  
### Dependencies

Include the following dependencies in each microservice's `pom.xml`:

1. **API Gateway**
    - Spring Boot Starter Web
    - Spring Cloud Gateway

2. **Aggregator, Customer, Product, and Order Services**
    - Spring Boot Starter Web
    - Spring Boot Starter Data JPA
    - H2 Database

#### Customer Service, Product Service, Order Service

1. Create new Spring Boot projects for each.
2. Set up JPA entities, Repositories, and Controllers as discussed in previous sections.

#### API Gateway

1. Create another Spring Boot project.
2. Add Spring Cloud Gateway dependency.
3. Configure routes in `application.yml`.

```yaml
spring:
  cloud:
    gateway:
      routes:
      - id: customer_route
        uri: http://localhost:8081
        predicates:
          - Path=/customers/**
      - id: product_route
        uri: http://localhost:8082
        predicates:
          - Path=/products/**
      - id: order_route
        uri: http://localhost:8083
        predicates:
          - Path=/orders/**
```

#### Aggregator

1. Create a Spring Boot project.
2. Use `RestTemplate` or `WebClient` to call other services.

### Step 2: Writing the Aggregator Logic

The Aggregator fetches data from Customer, Product, and Order services and aggregates it.

```java
@Service
public class AggregatorService {
    @Autowired
    private WebClient.Builder webClientBuilder;

    public AggregatedData getAggregatedData(Long customerId) {
        Customer customer = webClientBuilder.build()
            .get()
            .uri("http://localhost:8081/customers/" + customerId)
            .retrieve()
            .bodyToMono(Customer.class)
            .block();
        
        // Do the same for Product and Order and combine
        return new AggregatedData(customer, product, order);
    }
}
```

### Step 3: Testing

Use Postman or another API testing tool to test each service individually and then the aggregator.

## Working of Aggregator Pattern

- Fetches required data from individual services.
- Reduces the number of calls that a client has to make.
- Orchestrates and prepares a combined view of data.

## Best Practices, Do's and Don'ts, and Optimization Strategies

1. **Do's**
    - Always make services stateless.
    - Make services loosely coupled.
    - Always maintain a central, version-controlled repository for configurations.

2. **Don'ts**
    - Don't make direct calls from one microservice to another, always use the aggregator.
    - Don't hardcode configurations, use centralized configuration management.

3. **Optimization Strategies**
    - Use caching in the aggregator for frequent queries.
    - Make the aggregator as lightweight as possible.

4. **Best Practices**
    - Implement API throttling in API Gateway.
    - Use Circuit Breaker patterns.

This should provide a fairly detailed, hands-on example to help understand the ins and outs of using API Gateway and Aggregator patterns in a Spring-based microservices environment.

---

## Code Repository Structure and Other Details

Maintaining a well-organized code repository is essential for effective development, especially in microservices architecture where multiple projects may be involved.

Here is how the repository might look like:

```
- e-commerce-microservices/
  - api-gateway/
    - src/
      - main/
        - java/
          - com/
            - ecommerce/
              - gateway/
                - ApiGatewayApplication.java
                - config/
                  - RouteConfig.java
    - pom.xml
  - aggregator-service/
    - src/
      - main/
        - java/
          - com/
            - ecommerce/
              - aggregator/
                - AggregatorServiceApplication.java
                - service/
                  - AggregatorService.java
    - pom.xml
  - customer-service/
    ...
  - product-service/
    ...
  - order-service/
    ...
  - README.md
```

#### Versioning

Version control is essential. Each microservice should be versioned separately, with Git tags indicating releases.

#### Database Migrations

Maintain database schema changes in version control and automate database migrations using tools like Flyway or Liquibase.

#### Continuous Integration/Continuous Deployment (CI/CD)

Automate building, testing, and deployment of each microservice using CI/CD pipelines.

### Code Quality

Ensure high code quality using static code analysis tools like SonarQube.

### Testing

Write Unit Tests, Integration Tests, and End-to-End tests to ensure that your services are working as expected.

#### Unit Tests

Focus on testing each component or functionality of a single service. Use JUnit for this.

#### Integration Tests

Test the service as a whole and also test its dependencies on other services and databases. For this, you may employ Spring Boot's Test capabilities.

#### End-to-End Tests

Test the entire flow of an application from start to finish. Ensure that the entire process of a user accessing and using the site works smoothly.

### Final Words

It's essential to document each microservice's responsibility, data models, API documentation, etc., for better collaboration among team members. Always keep in mind the best practices, do's and don'ts, and optimization strategies when developing microservices.

---

Let's dive deeper into the details of each microservice involved in the e-commerce application.

#### Responsibilities
- Route incoming requests to appropriate microservices.
- Perform API composition for certain types of requests.

#### Configuration & Setup
- Include Spring Cloud Gateway as a dependency.
- Define the routes for different microservices in `application.yml`.

##### Code Hints
To set up route configurations, you can use `application.yml` or directly use Java configuration using `RouteLocator` bean.

```java
@Bean
public RouteLocator customRouteLocator(RouteLocatorBuilder builder) {
    return builder.routes()
        .route("customer_route", r -> r.path("/customers/**")
                .uri("http://localhost:8081"))
        .build();
}
```

#### Differences from other services
- Doesn't have its own data model or database.
- May include API rate-limiting, caching, etc.

---

#### Responsibilities
- Fetch and aggregate data from various services.
- Respond to requests from the API gateway.

#### Configuration & Setup
- Include WebClient to make async HTTP requests.
- Set up models to hold aggregated data.

##### Code Hints
Example using `WebClient` to fetch data asynchronously:

```java
public Mono<AggregatedResponse> aggregateData(Long customerId) {
    Mono<Customer> customer = webClientBuilder.build().get().uri("http://localhost:8081/customers/" + customerId).retrieve().bodyToMono(Customer.class);
    // Fetch products and orders similarly and aggregate
}
```

#### Differences from other services
- Doesn't have its own database.
- Solely responsible for orchestrating data and not managing it.

---

#### Responsibilities
- CRUD operations for customer data.
- Respond to requests from the aggregator or API gateway.

#### Configuration & Setup
- Set up a new Spring Boot project.
- Include Spring Data JPA and H2 database for storage.

##### Code Hints
Entity Class Example:

```java
@Entity
public class Customer {
    @Id
    @GeneratedValue
    private Long id;
    private String name;
    // getters and setters
}
```

Repository Interface:

```java
public interface CustomerRepository extends JpaRepository<Customer, Long> {}
```

#### Differences from other services
- Manages Customer data.
- Has its own database.

---

### 4. Product Service and Order Service

Follow similar guidelines to Customer Service, but the data models will be products and orders, respectively.

---

By understanding each service in-depth, you'll get a better understanding of how they work independently and yet function as a unified system. Each service encapsulates a specific business logic and data and interacts with other services to provide a complete end-to-end solution.

---

Let's elaborate on how these microservices connect with each other and demonstrate some workflows.

### Connection Mechanisms

1. **API Gateway** serves as the main entry point for all external requests and routes them to the respective microservice. 
2. **Aggregator Service** communicates with other services like Customer, Product, and Order services using RESTful APIs to fetch data.
3. The individual **CRUD Services (Customer, Product, Order)** don't talk to each other directly but communicate via the Aggregator or the API Gateway.

#### 1. Customer Makes a New Order

1. **Frontend**: User selects a product and places an order.
2. **API Gateway**: Receives the request and routes it to the Aggregator Service.
3. **Aggregator Service**: 
    - Calls the Product Service to verify if the product is in stock.
    - Calls the Customer Service to fetch customer information.
    - Finally, instructs the Order Service to create a new order.
4. **Order Service**: Creates a new order and returns the order details.
5. **Aggregator Service**: Aggregates the data and sends it back through the API Gateway.
6. **API Gateway**: Returns the data back to the frontend.

#### 2. User Wants to See their Profile Along With Past Orders

1. **Frontend**: User navigates to the profile section.
2. **API Gateway**: Routes the profile request to Aggregator Service.
3. **Aggregator Service**:
    - Fetches customer details from Customer Service.
    - Fetches all past orders related to the customer from Order Service.
4. **Aggregator Service**: Aggregates the customer and order details and sends it back.
5. **API Gateway**: Returns the data to the frontend.

#### API Gateway

You could set up dynamic routing or add filters for pre-processing and post-processing requests. For example, logging incoming requests.

```java
public class CustomGlobalFilter implements GlobalFilter {
    @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        // Log the request
        return chain.filter(exchange);
    }
}
```

#### Aggregator Service

To make multiple calls asynchronously and to improve performance, you can use `WebClient` like so:

```java
Mono<Product> product = webClient.get().uri("http://product-service/products/{id}", productId).retrieve().bodyToMono(Product.class);
Mono<Customer> customer = webClient.get().uri("http://customer-service/customers/{id}", customerId).retrieve().bodyToMono(Customer.class);

return Mono.zip(product, customer)
    .map(tuple -> {
        // Create aggregated object here
    });
```

#### Customer, Product, and Order Services

These services would usually consist of a Controller, Service, Repository, and Entity. They will have standard CRUD operations and would be implemented using Spring Boot and Spring Data JPA.

For example, the Customer Service might look like:

```java
@RestController
@RequestMapping("/customers")
public class CustomerController {
    @Autowired
    private CustomerService service;

    @GetMapping("/{id}")
    public ResponseEntity<Customer> getCustomer(@PathVariable Long id) {
        return ResponseEntity.ok(service.findById(id));
    }
    // more CRUD methods
}
```

These are just basic examples, but they should provide you with the idea of how individual microservices are designed, what their responsibilities are, and how they connect with each other. In the grand scheme, each microservice is an isolated piece of business functionality that collaborates with others to provide an end-to-end feature for the user.

---

### Extended Implementation Details

Let's consider the nuances that could be involved in each of these services to make them more resilient, efficient, and capable of handling real-world scenarios.

#### API Gateway

- **Load Balancing**: Use Spring Cloud LoadBalancer or Ribbon for client-side load balancing.

```yml
spring:
  cloud:
    loadbalancer:
      ribbon:
        enabled: true
```

- **Circuit Breaking**: Implement a circuit breaker like Resilience4J or Hystrix to handle failures in downstream services.

```java
@CircuitBreaker(name = "default", fallbackMethod = "fallback")
public Mono<MyResponse> someFunction() {
    // call downstream services
}
```

#### Aggregator Service

- **Cache**: Consider implementing a caching layer to store the frequently aggregated data.

```java
@Cacheable(value = "aggregatedData", key = "#customerId")
public AggregatedData getAggregatedData(Long customerId) {
    // aggregation logic
}
```

- **Asynchronous Operations**: To improve performance, you could fetch data from multiple services in parallel using `WebClient`.

```java
Mono.zip(
    webClient.get().uri("http://product-service/products").retrieve().bodyToMono(Product[].class),
    webClient.get().uri("http://order-service/orders").retrieve().bodyToMono(Order[].class)
).flatMap(tuple2 -> {
    // Aggregate the data and create the response object
});
```

#### Customer, Product, and Order Services

- **Validation**: Perform input validation using annotations or custom validators.

```java
public class Customer {
    @NotNull
    @Size(min = 1, max = 50)
    private String name;
    // ...
}
```

- **Auditing**: Use Spring Data's auditing capabilities to automatically fill in metadata like `createdBy`, `createdDate`, etc.

```java
@EntityListeners(AuditingEntityListener.class)
public class Customer {
    @CreatedBy
    private String createdBy;

    @CreatedDate
    private LocalDateTime createdDate;
    // ...
}
```

### Inter-service Communication in the Workflow

In a typical e-commerce application, orchestration might be preferred over choreography due to its straightforward nature and easier debugging. Here, the **Aggregator Service** could act as the orchestrator that coordinates various actions among the microservices.

- **Step-by-Step Orchestration**: For complex operations, you could break down the business logic into multiple steps and let the Aggregator Service guide each microservice through each step.

For instance, when placing an order:
1. **Aggregator Service** first validates the request and checks for stock by contacting **Product Service**.
2. Then it proceeds to reserve the stock.
3. It fetches **Customer** data and validates payment information.
4. Finally, it communicates with the **Order Service** to complete the order placement.

### Security Measures

- **OAuth2 or JWT** for secure communication between services.
- **HTTPS** for all service-to-service and external-to-service communications.

### Observability

1. **Logging**: Consistent logging across services. 
2. **Monitoring**: Prometheus and Grafana for real-time metrics.
3. **Tracing**: Jaeger or Zipkin for distributed tracing.

With these advanced details and considerations, your microservices would not only be well-structured but also robust and scalable, capable of handling real-world complexities.

**Comment by singhRatan777** [comment]
[api-gateway with aggregator.zip](https://github.com/akash-coded/spring-framework/files/12509346/api-gateway.with.aggregator.zip)
Ratan Singh Dewada

**Comment by AnuragMishra58** [comment]
Hi @akash-coded 
Please find below apps for aggregated pattern with API Gateway
Name: Anurag Mishra
Emp Id: 550966
[api-gateway-with-aggregator.zip](https://github.com/akash-coded/spring-framework/files/12511548/api-gateway-with-aggregator.zip)

**Reply by AnuragMishra58** [reply]
Added  OpenAPI 3/Swagger UI for all the services.
Accessible at: http://localhost:8081/swagger-ui/index.html
[api-gateway-with-aggregator-swagger-openapi.zip](https://github.com/akash-coded/spring-framework/files/12512060/api-gateway-with-aggregator-swagger-openapi.zip)

**Comment by jay4tech** [comment]
Hi @akash-coded

Please find below apps for mid-assessment with API Gateway
Name: Jay Prakash Kumar
Swagger URI -  https://<server:port>/swagger-ui


https://github.com/jay4tech/customer-service
https://github.com/jay4tech/product-service-api-gateway
https://github.com/jay4tech/order-service-api-gateway
https://github.com/jay4tech/discovery-service
https://github.com/jay4tech/api-gateway
https://github.com/jay4tech/aggregator-service-api-gateway

**Comment by HarshDPandya** [comment]
Hi @akash-coded , please find attached zip file for **Microservices - Design Patterns - API Gateway with Aggregator**
**Mid-Assessment includes (Microservice Architecture with API Gateway, Aggregator, Eureka Discovery Server, OpenFeign Client, Swagger and H2 database)**
Name: Harsh Pandya
TCS Emp Id: 2066481
[microservice-apigateway-aggregator-eurekaserver-2066481.zip](https://github.com/akash-coded/spring-framework/files/12513805/microservice-apigateway-aggregator-eurekaserver-2066481.zip)

**Comment by sagarrane91** [comment]
Hi @akash-coded ,
PFA Mid-Assessment
Sagar Rane
[Mid-Assessment.zip](https://github.com/akash-coded/spring-framework/files/12515055/Mid-Assessment.zip)

**Comment by zafvirinchi** [comment]
[Services.zip](https://github.com/akash-coded/spring-framework/files/12515618/Services.zip)
[gateway-service.zip](https://github.com/akash-coded/spring-framework/files/12515619/gateway-service.zip)
[aggregator-service.zip](https://github.com/akash-coded/spring-framework/files/12515620/aggregator-service.zip)
Hi @akash-coded  **Please find the attached projects**

**Comment by Prasadpoojary** [comment]
Hi 
Please find below,
Mid-Assessment includes (Microservice Architecture with API Gateway, Aggregator, Eureka service discovery, OpenFeign Client). 
Name : Prasad

[API-Gateway with Eureka and Aggregator.zip](https://github.com/akash-coded/spring-framework/files/12516327/API-Gateway.with.Eureka.and.Aggregator.zip)

Thanks,

**Comment by aathira-ps** [comment]
Hi Akash,

PFA for the mid assessment
[mid assesment.zip](https://github.com/akash-coded/spring-framework/files/12517595/mid.assesment.zip)
 projects

**Comment by naikkushal** [comment]
Hi Akash
Mid Assessment (Kushal Naik)
[Kushal_Naik_Mid_Assessment.zip](https://github.com/akash-coded/spring-framework/files/12517634/Kushal_Naik_Mid_Assessment.zip)

**Comment by elcenninan** [comment]
[aggregatorwithAPIgateway.zip](https://github.com/akash-coded/spring-framework/files/12509416/aggregatorwithAPIgateway.zip)
Elcen Zachariah-elcen.z@tcs.com

**Comment by arupmukherjee38** [comment]
[microservice_4_mid.zip](https://github.com/akash-coded/spring-framework/files/12518037/microservice_4_mid.zip)
Arup Mukherjee (Tcs emp id:1524555)

**Comment by Hemaruban** [comment]
Hi Akash,
PFA for assessment (Hemalatha R - 1085177)
[158-Hemalatha-API Gateway with Aggregator.zip](https://github.com/akash-coded/spring-framework/files/12518171/158-Hemalatha-API.Gateway.with.Aggregator.zip)

**Comment by Nagaraj2610** [comment]
Nagaraj
[assessment.zip](https://github.com/akash-coded/spring-framework/files/12518173/assessment.zip)
.

**Comment by arfaz9030** [comment]
Mohammad Arfaz(2354634)
https://github.com/arfaz9030/ecommerce-aggregate-microservices/tree/main

**Comment by Chandrika1999** [comment]
Hi
Please find below,
[microservice-apigateway-aggregator-eurekaserver.zip](https://github.com/akash-coded/spring-framework/files/12523512/microservice-apigateway-aggregator-eurekaserver.zip)
Mid-Assessment includes (Microservice Architecture with API Gateway, Aggregator, Eureka service discovery, OpenFeign Client).
Name : Chandrika Cheerla

**Comment by Devipadala** [comment]
Hi
Please find below,
[microservice-MidAssessment.zip](https://github.com/akash-coded/spring-framework/files/12523572/microservice-MidAssessment.zip)
Mid-Assessment includes (Microservice Architecture with API Gateway, Aggregator, Eureka service discovery, OpenFeign Client).
Name : Devi Padala

**Comment by arunsankar172** [comment]
Hi Akash
Arunsankar R
[Arunsankar_API Gateway & Aggregator Pattern.zip](https://github.com/akash-coded/spring-framework/files/12536377/Arunsankar_API.Gateway.Aggregator.Pattern.zip)

**Comment by esubshaik1610** [comment]
hi Akash,
Esub  shaik[2268851]
[session6.zip](https://github.com/akash-coded/spring-framework/files/12537166/session6.zip)

**Comment by Kiran-Ramesh1** [comment]
Hi @akash-coded ,

Name : Kiran Ramesh (1949256)
[Aggregator Using ApiGateway.zip](https://github.com/akash-coded/spring-framework/files/12568622/Aggregator.Using.ApiGateway.zip)

PFA for assessment

**Comment by jay4tech** [comment]
Hi @akash-coded 
Name: Jay Prakash Kumar

https://github.com/jay4tech/api-gateway/tree/msa_158
https://github.com/jay4tech/aggregator-service-api-gateway/tree/msa_158
https://github.com/jay4tech/customer-service/tree/msa_158
https://github.com/jay4tech/product-service-api-gateway/tree/msa_158
https://github.com/jay4tech/order-service-api-gateway/tree/msa_158

Swagger UI -
http://localhost:8083/swagger-ui/index.html -- Order Service
http://localhost:8082/swagger-ui/index.html -- Product Service 
http://localhost:8081/swagger-ui/index.html -- Customer Service
http://localhost:9090/swagger-ui/index.html -- Aggregate Service

API Gateway - 
http://localhost:8080/aggregate/orders/1  -- get an aggregated response from aggregate service using different services

**How to test -**

 **Create customer**
`curl -X 'POST' \
  'http://localhost:8081/customers/' \
  -H 'accept: */*' \
  -H 'Content-Type: application/json' \
  -d '{
  "name": "Jay Prakash",
  "email": "info@professionalhacker.in",
  "password": "IAMNOTGONATELLYOU"
}'`

 **Create Product**

`curl -X 'POST' \
  'http://localhost:8082/products/' \
  -H 'accept: */*' \
  -H 'Content-Type: application/json' \
  -d '{
  "name": "iPhone 15",
  "price": 150000
}'`

 **Create Order**

`curl -X 'POST' \
  'http://localhost:8083/orders/' \
  -H 'accept: */*' \
  -H 'Content-Type: application/json' \
  -d '{
  "product": "iPhone 15",
  "customerId": 1
}'`

** Call API Gateway to get aggregated response from aggregate service using different services**
`curl -X 'GET' \
  'http://localhost:8080/aggregate/orders/1' \
  -H 'accept: */*'`

**Comment by shriramram** [comment]
@akash-coded 
Name - Dattatraya Shriram
Description - Aggregator assignment implementation
[microservices-aggregator.zip](https://github.com/akash-coded/spring-framework/files/12806668/microservices-aggregator.zip)

**Comment by arghyagiri** [comment]
@akash-coded : https://github.com/arghyagiri/microservice-e2/blob/main/e-commerce-microservices/Read.md

**Comment by Rajashree-Ravi** [comment]
@akash-coded 

Please find attached assignment implementation  for the combination pattern (Aggregated & API gateway) 

[api-gateway-and-aggregator-pattern.zip](https://github.com/akash-coded/spring-framework/files/12822779/api-gateway-and-aggregator-pattern.zip)
